### PR TITLE
Support materialized view load and insert

### LINKED ISSUE #3344: [Proposal] Materialized View 2.0
# The status of materialized view 1.0

In the present, we have supported the materialized views in Doris 0.12 version. The materialized view selector supports to select the most efficient mv and rewrite the SQL to query against the selected mv instead of the base table.
For query results contain a small number of rows where the original table has a large amount of data, the performance can reach the 5X to 100X times depends on the cardinality of the data.
The aggregate functions supported by the materialized view in 0.12 include: sum, min, max.

However, the aggregate functions supported by the current materialized view are not rich enough to fully cover the user's scene.
For example, in the `Order` scenario, user needs to analyze the number of orders in different dimensions.
Another example is the `count_distinct` function is used for analyzing PV and UV data in website traffic.

# The goal of materialized view 2.0

In order to support more scenarios, the materialized view 2.0 will support the following functions:

1. Materialized view supports aggregate functions: count, count_distinct (bitmap and hll)
2. Support to create materialized views of the same column with different aggregate functions. For example: ```select k1, sum (v1), min (v1) from table group by k1```

### PR DESCRIPTION
**Pull request by HangyuanLiu**
This commit mainly supports load bitmap_union, hll_union, and count materialized views.

The main changes are as follows:
1、insert stmt support load extend column
2、load stmt support load extend column

Issue : #3344

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/load/Load.java [hunk 1]
```diff
@@ -1013,6 +1013,13 @@ public static void initColumns(Table tbl, List<ImportColumnDesc> columnExprs,
                 slotDescByName.put(realColName, slotDesc);
             }
         }
+        // The extension column of the materialized view is added to the expression evaluation of load
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/load/Load.java:1016]: Please add some example

File: fe/src/main/java/org/apache/doris/load/Load.java [hunk 2]
```diff
@@ -1013,6 +1013,13 @@ public static void initColumns(Table tbl, List<ImportColumnDesc> columnExprs,
...
+        for (Column column : tbl.getFullSchema()) {
+            if (column.getDefineExpr() != null) {
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/load/Load.java:1018]: Are you sure that this is correct` column(a, tmp_c) set(b=f(c), c=b) `？

**Review by EmmyMiao87 — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/analysis/InsertStmt.java [hunk 3]
```diff
@@ -453,6 +453,26 @@ private void analyzeSubquery(Analyzer analyzer) throws UserException {
...
+            Column column = targetTable.getFullSchema().get(mvColumnIdx);
+            if (column.isNameWithPrefix(CreateMaterializedViewStmt.MATERIALIZED_VIEW_NAME_PRFIX)) {
+                List<Expr> slots = new ArrayList<>();
+                column.getDefineExpr().collect(SlotRef.class, slots);
+                //We only support one children of define expr in materialized view column
+                Preconditions.checkArgument(slots.size() == 1);
+                String origName = ((SlotRef) slots.get(0)).getColumnName();
+                for (int originColumnIdx = 0; originColumnIdx < targetColumns.size(); originColumnIdx++) {
+                    if (targetColumns.get(originColumnIdx).nameEquals(origName, false)) {
+                        origColIdx2MvColIdx.put(mvColumnIdx, originColumnIdx);
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/analysis/InsertStmt.java:468]: mvColumnIdxToOrigColumnIdx?

File: fe/src/main/java/org/apache/doris/analysis/InsertStmt.java [hunk 4]
```diff
@@ -519,17 +556,23 @@ private void analyzeSubquery(Analyzer analyzer) throws UserException {
         }
 
         // expand baseTblResultExprs and colLabels in QueryStmt
-        if (!origColIdxsForShadowCols.isEmpty()) {
+        if (!origColIdxsForShadowCols.isEmpty() || !origColIdx2MvColIdx.isEmpty()) {
             if (queryStmt.getResultExprs().size() != queryStmt.getBaseTblResultExprs().size()) {
                 for (Integer idx : origColIdxsForShadowCols) {
                     queryStmt.getBaseTblResultExprs().add(queryStmt.getBaseTblResultExprs().get(idx));
                 }
+                for (Integer idx : origColIdx2MvColIdx.keySet()) {
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/analysis/InsertStmt.java:564]: The key of origColIdx2MvColIdx is mvColIdx. The mvColumn has been supplied at the end of result exprs in query stmt. So I think it is not match.

**Review by EmmyMiao87 — CHANGES_REQUESTED** [review]

**Comment by EmmyMiao87** [comment]
Is it possible to unify the columns of schema change and materialized view into one logical process in insert stmt?

File: fe/src/main/java/org/apache/doris/analysis/InsertStmt.java [hunk 5]
```diff
@@ -453,6 +453,26 @@ private void analyzeSubquery(Analyzer analyzer) throws UserException {
             }
         }
 
+        //Get the correspondence of this column to the original column through defineExpr
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/analysis/InsertStmt.java:456]: Please add example of mv column and schema change column

File: fe/src/main/java/org/apache/doris/load/Load.java (hunk 2 above)
Comment (HangyuanLiu) [reply fe/src/main/java/org/apache/doris/load/Load.java:1024]: > Are you sure that this is correct`column(a, tmp_c) set(b=f(c), c=b)`？

Fix done

**Review by morningman — DISMISSED** [review]
LGTM

File: fe/src/main/java/org/apache/doris/analysis/InsertStmt.java [hunk 6]
```diff
@@ -521,22 +530,22 @@ private void analyzeSubquery(Analyzer analyzer) throws UserException {
...
                     ExprSubstitutionMap smap = new ExprSubstitutionMap();
-                    smap.getLhs().add(slots.get(0));
-                    smap.getRhs().add(queryStmt.getResultExprs().get(value));
+                    smap.getLhs().add(mvColumn.getRefColumn());
+                    smap.getRhs().add(queryStmt.getResultExprs().get(origColIdx));
 
-                    queryStmt.getResultExprs().add(Expr.substituteList(Lists.newArrayList(expr), smap, analyzer, false).get(0));
+                    Expr e = Expr.substituteList(Lists.newArrayList(mvColumn.getDefineExpr()), smap, analyzer, false).get(0);
+                    queryStmt.getResultExprs().add(e);
+                    queryStmt.getBaseTblResultExprs().add(e);
```
Comment (EmmyMiao87) [inline fe/src/main/java/org/apache/doris/analysis/InsertStmt.java:548]: What's different between `resultExpr` and `baseTblResultExpr`

**Review by EmmyMiao87 — CHANGES_REQUESTED** [review]

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**